### ISSUE TITLE
rpath error on Catalina, symptom: bzip2 doesn't work

### ISSUE THREAD

**Issue by healther** [spack/spack#13245]
When rebuilding my software stack on Catalina I encountered the following build error on Catalina
```
==> Installing openmpi
==> Searching for binary cache of openmpi
==> No binary for openmpi found: installing from source
==> Fetching http://www.open-mpi.org/software/ompi/v3.1/downloads/openmpi-3.1.4.tar.bz2
##################################################################################################################################################################################################### 100.0%##################################################################################################################################################################################################### 100.0%##################################################################################################################################################################################################### 100.0%
==> Staging archive: /var/folders/1q/jxx1cgcx3l13d153s6_c1zc00000gn/T/healther/spack-stage/openmpi-3.1.4-qecenftbojb2tzv6zrtsz64nzenb4h2x/openmpi-3.1.4.tar.bz2
dyld: Library not loaded: @rpath/libbz2.1.dylib
  Referenced from: /Users/healther/Documents/git.nosync/spack/var/spack/environments/mypython/.spack-env/view/bin/bzip2
  Reason: image not found
/Users/healther/Documents/git.nosync/spack/var/spack/environments/mypython/.spack-env/view/bin/tar: Child died with signal 6
/Users/healther/Documents/git.nosync/spack/var/spack/environments/mypython/.spack-env/view/bin/tar: Error is not recoverable: exiting now
==> Error: ProcessError: Command exited with status 2:
    '/Users/healther/Documents/git.nosync/spack/var/spack/environments/mypython/.spack-env/view/bin/tar' '-xf' '/var/folders/1q/jxx1cgcx3l13d153s6_c1zc00000gn/T/healther/spack-stage/openmpi-3.1.4-qecenftbojb2tzv6zrtsz64nzenb4h2x/openmpi-3.1.4.tar.bz2'
```

The problem originates in the `bzip2` binary installed by spack. Calling it even without arguments leads to
```
% /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/bin/bzip2

dyld: Library not loaded: @rpath/libbz2.1.dylib
  Referenced from: /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/bin/bzip2
  Reason: image not found
zsh: abort
```

Reproduction should be possible by doing `spack install bzip2` and calling the installed binary on any OSX Catalina system.

@adamjstewart I assume you also rebuild your software stack but somehow avoided that error (or did I miss an issue?)

**Comment by adamjstewart** [comment]
@healther I did not encounter this issue. Can you post the output of:
```console
$ otool -L /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/lib/libbz2.dylib
```
For the record, when I run the `bzip2` executable Spack installs, it successfully finds its libraries and doesn't crash like you showed.

**Comment by healther** [comment]
```
% otool -L /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/lib/libbz2.dylib

/Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/lib/libbz2.dylib:
	@rpath/libbz2.1.dylib (compatibility version 1.0.8, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
```
Not that I know what any of that means^^

> For the record, when I run the bzip2 executable Spack installs, it successfully finds its libraries and doesn't crash like you showed.

Mh that is interesting. Are you using `zsh` or `bash`? That's about the only thing where I would expect anything special to have come from. Otherwise I'm running a completely standard `spack/develop:HEAD` and (for somewhat unrelated reasons) I did a `git clean -dfx` and a `rm -r ~/.spack`, so there _should_ have been no state surviving...
That being said, I have added `gfortran` from a `gcc@9.2.0` to my native clang compiler to be able to build `blas`

**Comment by adamjstewart** [comment]
My `otool` output looks the same, with the `@rpath/libbz2.1.dylib`. I'm still using Bash, and I also added `gfortran` from `gcc@9.2.0` to my native clang compiler. @gartung can you take a look at this issue? You probably know more than I do about macOS linking.

**Comment by gartung** [comment]
Try 'otool -l | grep -A3 LC_RPATH' and see if the rpath is set correctly.

**Comment by healther** [comment]
Your command comes back completely empty, just taking all `LC`s looks like:
```
% otool -l /Users/healther/Documents/git.nosync/spack/var/spack/environments/mypython/.spack-env/view/lib/libbz2.dylib | grep LC
      cmd LC_SEGMENT_64
      cmd LC_SEGMENT_64
      cmd LC_SEGMENT_64
      cmd LC_SEGMENT_64
          cmd LC_ID_DYLIB
            cmd LC_DYLD_INFO_ONLY
     cmd LC_SYMTAB
            cmd LC_DYSYMTAB
     cmd LC_UUID
       cmd LC_BUILD_VERSION
      cmd LC_SOURCE_VERSION
          cmd LC_LOAD_DYLIB
      cmd LC_FUNCTION_STARTS
      cmd LC_DATA_IN_CODE
```

**Comment by gartung** [comment]
So 'otool -D' should show that the library ID used when linking with other binaries is @rpath/libbz2.dylib. The next step is to check the output of 'otool -L bzip2' and make sure the LC_RPATH field has path set to the install prefix.

**Comment by healther** [comment]
```
healther@dhcp647-117 spack % otool -D /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/lib/libbz2.dylib 
/Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/lib/libbz2.dylib:
@rpath/libbz2.1.dylib
healther@dhcp647-117 spack % otool -L /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/bin/bzip2
/Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-broadwell/clang-11.0.0-apple/bzip2-1.0.8-fodftvdvnzr3fcj3cuwqhidrq3dnggb6/bin/bzip2:
	@rpath/libbz2.1.dylib (compatibility version 1.0.8, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
```

**Comment by gartung** [comment]
Sorry I mean 
'otool -l bzip2'

**Comment by healther** [comment]
Assuming that this was a capital i it returned a longer output, see 

https://gist.github.com/healther/4024695ecc3fe3222408bdcde97a9167

with (hopefully all outputs)

**Comment by gartung** [comment]
Not capital I lower case L. See man otool

**Comment by gartung** [comment]
My guess is that LC_RPATH was not set or set incorrectly.

**Comment by healther** [comment]
I'm doing `otool -l <path/to/spack/opt/.../bzip2/bin/bzip2` (lowercase L) as I don't have the `bzip2` binary on my global `PATH`, I assume that is still okay. There is no `LC_RPATH` entry at all. Would this show up in the build-log?

In there I find
```
Doing 6 tests (3 compress, 3 uncompress) ...
If there's a problem, things might stop at this point.
 
./bzip2 -1  < sample1.ref > sample1.rb2
./bzip2 -2  < sample2.ref > sample2.rb2
./bzip2 -3  < sample3.ref > sample3.rb2
./bzip2 -d  < sample1.bz2 > sample1.tst
./bzip2 -d  < sample2.bz2 > sample2.tst
./bzip2 -ds < sample3.bz2 > sample3.tst
cmp sample1.bz2 sample1.rb2 
cmp sample2.bz2 sample2.rb2
cmp sample3.bz2 sample3.rb2
cmp sample1.tst sample1.ref
cmp sample2.tst sample2.ref
cmp sample3.tst sample3.ref
```

so I assume things worked at some point during the build

**Comment by gartung** [comment]
Having no  LC_RPATH set in bzip2 would be the problem. @adamjstewart  is something done to bzip2 after the install?

**Comment by gartung** [comment]
@adamjstewart It looks like the bzip2 package will need a call to
install_name_tool -add-rpath <prefix>/lib  <prefix>/bin//bzip2
as a post install step.

**Comment by healther** [comment]
Mh okay, doing things the full spack-way, i.e. with `spack load -r bzip2` `bzip2` no longer errors out. 

Before I was using the view from my environment and I was putting `bin`, `lib`, `man` on `PATH`, `LIBRARY_PATH` and `MANPATH` (initially) or trying to use the binary `bzip2` directly without adding any of spack's directories to *PATH in my environment. I also didn't have `environment-modules` installed, I assume that not having this available is what triggered my original build error in `openmpi`

Questions:
1) should using the bare binary (without env modifications) work?
2) Is there a change in behaviour between Mojave/Catalina or is this only due to me lacking `environment-modules`?

**Comment by gartung** [comment]
If you have SIP enabled, which is the default, then DYLD_LIBRARY_PATH is unset. If the ancient bzip2 scripts relied on DYLD_LIBRARY_PATH to set LC_RPATH, then that might explain why LC_RPATH was not set in your bzip2 binary.

**Comment by gartung** [comment]
All of our macOS build machines have SIP disabled for this reason.

**Comment by gartung** [comment]
You can check if you have SIP enabled with
csrutil status

**Comment by adamjstewart** [comment]
That's so strange, I have an almost identical setup but I don't have any problems with bzip2:
```console
$ otool -L lib/libbz2.dylib 
lib/libbz2.dylib:
	@rpath/libbz2.1.dylib (compatibility version 1.0.8, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
$ otool -L bin/bzip2
bin/bzip2:
	@rpath/libbz2.1.dylib (compatibility version 1.0.8, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
$ otool -l lib/libbz2.dylib | grep -A3 LC_RPATH
          cmd LC_RPATH
      cmdsize 136
         path /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/clang-11.0.0-apple/bzip2-1.0.8-vxe7pwivj5a4it5upk47itvaqdgjacjh/lib (offset 12)
Load command 13
--
          cmd LC_RPATH
      cmdsize 136
         path /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/clang-11.0.0-apple/bzip2-1.0.8-vxe7pwivj5a4it5upk47itvaqdgjacjh/lib64 (offset 12)
Load command 14
$ otool -l bin/bzip2 | grep -A3 LC_RPATH
          cmd LC_RPATH
      cmdsize 136
         path /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/clang-11.0.0-apple/bzip2-1.0.8-vxe7pwivj5a4it5upk47itvaqdgjacjh/lib (offset 12)
Load command 16
--
          cmd LC_RPATH
      cmdsize 136
         path /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/clang-11.0.0-apple/bzip2-1.0.8-vxe7pwivj5a4it5upk47itvaqdgjacjh/lib64 (offset 12)
Load command 17
$ csrutil status
System Integrity Protection status: enabled.
```
I've been trying to add `fix_darwin_install_tool` to packages as I find them, but if someone wants to modify it to always replace `@rpath` with the actual path, I would be fine with looking for that too.

**Comment by healther** [comment]
```
spack % csrutil status
System Integrity Protection status: enabled.
```
But that was already in Mojave right? However there I most likely had `environment-modules` installed and the modules manually set `DYLD_LIBRARY_PATH`.

The `install_name_tool -add-rpath prefix/lib prefix/bin//bzip2` thing should then be done for all rpathed binaries on OSX right?

**Comment by adamjstewart** [comment]
I run it from the installation prefix. I have no special environment variables set, I don't use modules.

**Comment by healther** [comment]
@adamjstewart do you mind posting your `spack-build-out.txt` and `spack-build-env.txt`? I'd like to do a diff to mine

**Comment by adamjstewart** [comment]
@healther here you go: [spack-build-out.txt](https://github.com/spack/spack/files/3751852/spack-build-out.txt)

**Comment by adamjstewart** [comment]
We should really add post-installation unit tests for this package. At the bare minimum, just running `bzip2 --help` would have uncovered this issue. If we figure out what needs to be changed to fix your setup, we can also add this at the same time.

**Comment by healther** [comment]
First of all thanks to you guys for the help, but now I cannot reproduce this issue... 

I _may_ have started the initial install after updating the OS but before updating Xcode and/or having have spack with some old state. I'm not entirely sure what I need to do in order to clean all state of spack, but before the latest attempt I did
```
git clean -dfx
spack clean -a
<start new shell>
spack install gcc
<update compilers.yaml to include gfortran>
spack env activate <myenv>
spack install
```
That one worked. I still have an Xcode update outstanding which I'll most likely do over the weekend I'll keep an eye out, but I fear this one will be down to "not reproducible". I'll also read up on the documentation on what kind of state spack is holding, I assume that it is too much (see also #12733 ) and I ran into an artifact of either building against an old repository (i.e. state from before a `git pull`) or an old compiler/system libraries.

Either way I assume that this is a fairly narrow window to hit, so I'm fine with closing this issue. Thoughts?
